File: sport-log-server/src/model/metcon.rs
use chrono::NaiveDateTime;
use diesel_derive_enum::DbEnum;
use serde::{Deserialize, Serialize};

use sport_log_server_derive::{Create, Delete, GetAll, GetById, Update};

use crate::{
    model::{MovementId, MovementUnit, UserId},
    schema::{metcon, metcon_movement, metcon_session},
};

#[derive(DbEnum, Debug, Serialize, Deserialize)]
pub enum MetconType {
    Amrap,
    Emom,
    ForTime,
    Ladder,
}

pub type MetconId = i32;

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "metcon"]
pub struct Metcon {
    pub id: MetconId,
    pub user_id: UserId,
    pub name: Option<String>,
    pub metcon_type: MetconType,
    pub rounds: Option<i32>,
    pub timecap: Option<i32>,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "metcon"]
pub struct NewMetcon {
    pub user_id: UserId,
    pub name: Option<String>,
    pub metcon_type: MetconType,
    pub rounds: Option<i32>,
    pub timecap: Option<i32>,
}

pub type MetconMovementId = i32;

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "metcon_movement"]
pub struct MetconMovement {
    pub id: MetconMovementId,
    pub movement_id: MovementId,
    pub metcon_id: MetconId,
    pub count: i32,
    pub unit: MovementUnit,
    pub weight: Option<f32>,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "metcon_movement"]
pub struct NewMetconMovement {
    pub movement_id: MovementId,
    pub metcon_id: MetconId,
    pub count: i32,
    pub unit: MovementUnit,
    pub weight: Option<f32>,
}

pub type MetconSessionId = i32;

#[derive(
    Queryable, AsChangeset, Serialize, Deserialize, Debug, Create, GetById, GetAll, Update, Delete,
)]
#[table_name = "metcon_session"]
pub struct MetconSession {
    pub id: MetconSessionId,
    pub user_id: UserId,
    pub metcon_id: MetconId,
    pub datetime: NaiveDateTime,
    pub time: Option<i32>,
    pub rounds: Option<i32>,
    pub reps: Option<i32>,
    pub rx: bool,
    pub comments: Option<String>,
}

#[derive(Insertable, Serialize, Deserialize)]
#[table_name = "metcon_session"]
pub struct NewMetconSession {
    pub user_id: UserId,
    pub metcon_id: MetconId,
    pub datetime: NaiveDateTime,
    pub time: Option<i32>,
    pub rounds: Option<i32>,
    pub reps: Option<i32>,
    pub rx: bool,
    pub comments: Option<String>,
}

File: sport-log-client/lib/database/tables/metcon.dart
import 'package:fixnum/fixnum.dart';
import 'package:sport_log/database/table.dart';
import 'package:sport_log/models/metcon/all.dart';

class MetconTable extends Table<Metcon> {
  @override
  String get setupSql => '''
create table metcon (
    user_id integer,
    name text check (length(name) <= 80),
    metcon_type integer not null check (metcon_type between 0 and 2),
    rounds integer check (rounds >= 1),
    timecap integer check (timecap > 0), -- seconds
    description text,
    $idAndDeletedAndStatus
);
  ''';

  @override
  DbSerializer<Metcon> get serde => DbMetconSerializer();

  @override
  String get tableName => 'metcon';
}

class MetconMovementTable extends Table<MetconMovement> {
  @override
  DbSerializer<MetconMovement> get serde => DbMetconMovementSerializer();

  @override
  String get setupSql => '''
create table metcon_movement (
    metcon_id integer not null references metcon(id) on delete cascade,
    movement_id integer not null references movement(id) on delete no action,
    movement_number integer not null check (movement_number >= 0),
    count integer not null check (count >= 1),
    movement_unit integer not null check (movement_unit between 0 and 6),
    weight real check (weight > 0),
    $idAndDeletedAndStatus
);
  ''';

  @override
  String get tableName => 'metcon_movement';

  Future<void> setSynchronizedByMetcon(Int64 id) async {
    database.update(tableName, Table.synchronized,
        where: '${Keys.metconId} = ?', whereArgs: [id.toInt()]);
  }

  Future<List<MetconMovement>> getByMetcon(Int64 id) async {
    final result = await database.query(tableName,
        where: '${Keys.metconId} = ? AND ${Keys.deleted} = 0',
        whereArgs: [id.toInt()]);
    return result.map(serde.fromDbRecord).toList();
  }

  Future<void> deleteByMetcon(Int64 id) async {
    await database.update(tableName, {Keys.deleted: 1},
        where: '${Keys.deleted} = 0 AND ${Keys.metconId} = ?',
        whereArgs: [id.toInt()]);
  }
}

class MetconSessionTable extends Table<MetconSession> {
  @override
  DbSerializer<MetconSession> get serde => DbMetconSessionSerializer();

  @override
  String get setupSql => '''
create table metcon_session (
    user_id integer not null,
    metcon_id integer not null references metcon(id) on delete no action,
    datetime text not null default (datetime('now')),
    time integer check (time > 0), -- seconds
    rounds integer check (rounds >= 0),
    reps integer check (reps >= 0),
    rx integer not null default 1 check (rx in (0, 1)),
    comments text,
    $idAndDeletedAndStatus
);
  ''';

  @override
  String get tableName => 'metcon_session';

  Future<bool> existsByMetcon(Int64 id) async {
    return (await database.rawQuery('''select 1 from $tableName
            where ${Keys.metconId} = ${id.toInt()}
              and ${Keys.deleted} = 0''')).isNotEmpty;
  }
}
